src/math/web_mercator.rs

```diff
@@ -0,0 +1,146 @@
+//! Calculations with Web Mercator coordinates and tiles.
+
+use nalgebra::Vector2;
+use nav_types::WGS84;
+use std::f64::consts::{FRAC_1_PI, PI};
+
+/// 2.0 * E.powf(PI).arctan() - FRAC_PI_2;
+/// In degrees, it's 85.051129 (cf. Wikipedia)
+const LAT_BOUND_RAD: f64 = 1.484_422_229_745_332_444_394_989_579_450_34;
+
+/// LAT_BOUND_SIN = sin(LAT_BOUND_RAD)
+const LAT_BOUND_SIN: f64 = 0.996_272_076_220_749_980_279_833_835_083_99;
+
+const TWO_PI: f64 = 2.0 * PI;
+const FOUR_PI: f64 = 4.0 * PI;
+const FOUR_PI_INV: f64 = 0.25 * FRAC_1_PI;
+const TILE_SIZE: u32 = 256;
+
+/// The max zoom level is currently 23 because it makes implementation easier,
+/// but theoretically nothing stops us from going deeper.
+pub const MAX_ZOOM: u8 = 23;
+
+/// A Web Mercator coordinate. Essentially a position in a 2D map of the world.
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub struct WebMercatorCoord {
+    /// Implementation detail: This is normalized to [0, 1), so not zoom level 0.
+    /// This makes calculations a bit simpler.
+    xy: Vector2<f64>,
+}
+
+impl WebMercatorCoord {
+    /// Projects a lat/lng coordinate to Web Mercator.
+    ///
+    /// Equivalent to the formula on [Wikipedia](https://en.wikipedia.org/wiki/Web_Mercator_projection#Formulas).
+    /// If the latitude is outside `[-85.051129, 85.051129]`, it is clamped to that interval first.
+    pub fn from_lat_lng(lat_lng: &WGS84<f64>) -> Self {
+        // Implemented according to
+        // https://developers.google.com/maps/documentation/javascript/examples/map-coordinates?csw=1
+        // but clamping is done before the sin() operation.
+        let lat = nalgebra::clamp(lat_lng.latitude(), -LAT_BOUND_RAD, LAT_BOUND_RAD);
+        let sin_y = lat.sin();
+
+        let xy = Vector2::new(
+            0.5 + lat_lng.longitude() / TWO_PI,
+            0.5 - ((1.0 + sin_y) / (1.0 - sin_y)).ln() * FOUR_PI_INV,
+        );
+        Self { xy }
+    }
+
+    /// Convert the Web Mercator coordinate back to lat/lng.
+    ///
+    /// The altitude returned is always 0.
+    pub fn to_lat_lng(&self) -> WGS84<f64> {
+        let centered = self.xy - Vector2::new(0.5, 0.5);
+
+        let sin_term = (-centered.y * FOUR_PI).exp();
+
+        // Note that sin_term = -(2/(sin(y)-1)) - 1
```

Comment:
is sin_term the sin_y computed in line 60? maybe unify the syntax

Reply:
Oh true. Done